### ISSUE TITLE
Type error with discriminatedUnion when using dynamic amount of objects

### ISSUE THREAD

**Issue by ghostdevv** [colinhacks/zod#1181]
For a project I am working on we are using zod to type check user provided input at runtime, we have "sections" which are defined in a manifest. I need to have a discriminatedUnion by am getting a type error when doing the following:

```ts
const section = z.discriminatedUnion('type', [
  ...Object.values(manifest.sections).map((section) => section.zod),
])
```

The above code does work, and what is weirder is that when accessing the items individually it works:

```ts
const zodObjects = Object.values(manifest.sections).map((section) => section.zod)

const section = z.discriminatedUnion('type', [
  zodObjects[0],
  zodObjects[1],
  zodObjects[2],
  // ...
])
```

At a guess it's because it doesn't know that it's a array with multiple items but am unsure the best way of fixing this

**Comment by Aaronius** [comment]
This issue is probably the same as https://github.com/colinhacks/zod/issues/989, except this issue is actually open (I can't get anyone to re-open the other issue).

**Comment by ghostdevv** [comment]
Still relevant

**Comment by benjick** [comment]
🔨

**Comment by Aaronius** [comment]
🔔

**Comment by benjick** [comment]
Boop

**Comment by scotttrinh** [comment]
👋 

Yeah, the issue here is that we enforce at the type level that the array has at least two items (`[T, T, ...T[]]`), but the type of `Object.values` is a (potentially empty) array. You can get around this in a few ways:

1. Make a function that wraps your `Object.values` calls and uses a runtime check and assertion that the value has two or more values.
2. Use some type casting to assert that the `Object.values` returns `[T, T, ...T[]]` to tell the compiler that you are absolutely certain it contains at least two values of type `T`.
3. Throw caution to the wind and use `as any` #yolo

---

```ts
import { z } from "zod";

const manifest = {
    sections: [
        z.object({ type: z.literal("main") }),
        z.object({ type: z.literal("nav") }),
        z.object({ type: z.literal("footer") })
    ]
};

function atLeastTwo<T>(arr: T[]): [T, T, ...T[]] {
  if (arr.length < 2) {
    throw new Error("Instantiated with fewer than two items");
  }

  return arr as [T, T, ...T[]];
}

const section = z.discriminatedUnion("type", [
    ...(atLeastTwo(manifest.sections)),
]);
```

---

TL;DR: we want to enforce that the array you pass to `discriminatedUnion` (and a few other constructors!) has at least two values, and the only way to satisfy TypeScript is to actually ensure that statically.

**Comment by yamcodes** [comment]
Expanding on @scotttrinh's answer, I was able to resolve it in way (2), meaning using TypeScript;

```ts
import { z } from "zod";

type AtLeastTwo<T> = [T, T, ...T[]];

/**
 * Options (schemas) to pass into `z.discriminatedUnion`.
 * Must be at least two elements.
 * See https://github.com/colinhacks/zod/issues/1181 for why this is needed.
 */
type ZodDiscriminatedUnionOptions = AtLeastTwo<z.ZodObject<z.ZodRawShape>>;

const schemas = [
  z.object({ type: z.literal("main") }),
  z.object({ type: z.literal("nav") }),
  z.object({ type: z.literal("footer") })
] satisfies ZodDiscriminatedUnionOptions;

const section = z.discriminatedUnion("type", schemas);
```

**Comment by yamcodes** [comment]
Expanding on @scotttrinh’s answer, we can do this with [Assertions (introduced in TypeScript 3.7)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions)

```ts
import { z } from "zod";

/**
 * An array with more than one element.
 */
export type MultipleElementsArray<T> = [T, T, ...T[]];

export function assertMultipleElementsArray<T>(
  arr: T[],
): asserts arr is MultipleElementsArray<T> {
  if (arr.length < 2) throw new Error("Instantiated with fewer than two items");
}

assertMultipleElementsArray(schemas);

const section = z.discriminatedUnion("type", schemas);
```

Whereas schemas is an array/tuple of `z.Object`s with the `discriminator` as an entry (like `manifest.sections` in your example)

**Comment by ffraenz** [comment]
With both `atLeastTwo` and `assertMultipleElementsArray` I encountered the issue that `T` is widening over all types of elements of the array. With that, a lot of type information is potentially lost.

Here's my example:

```ts
export const emailTemplates = [
  { name: 'email1', propsSchema: z.object({ ... }), render: props => ... },
  { name: 'email2', propsSchema: z.object({ ... }), render: props => ... },
  { name: 'email3', propsSchema: z.object({ ... }), render: props => ... },
] as const

export const sendEmailArgsSchema =
  z.discriminatedUnion(
    'template', 
    atLeastTwo(emailTemplates.map(emailTemplate =>
      z.object({
        to: z.string(),
        template: z.literal(emailTemplate.name),
        props: emailTemplate.propsSchema
      })
    ))
  )

type SendEmailArgs = z.infer<typeof sendEmailArgsSchema>
```

Here, the intended discriminated union `SendEmailArgs` is widened into a single type:

```
{
    to: string;
    template: "email1" | "email2" | "email3";
    props: {
        ...;
    } | ... 1 more ... | {
        ...;
    };
}
```

In practice, I can send `email2` props with the `email1` template and no issue is raised by TypeScript.
